cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt

```diff
@@ -0,0 +1,301 @@
+ * are new nodes or because a new lattice is bigger than the old one).
+ */
+open class State<K, V> : IdentityHashMap<K, Lattice<V>>() {
+
+    /**
+     * It updates this state by adding all new nodes in [other] to `this` and by computing the least
+     * upper bound for each entry.
+     *
+     * Returns this and a flag which states if there was any update necessary (or if `this` is equal
+     * before and after running the method).
+     */
+    open fun lub(other: State<K, V>): Pair<State<K, V>, Boolean> {
+        var update = false
+        for ((node, newLattice) in other) {
+            update = push(node, newLattice) || update
+        }
+        return Pair(this, update)
+    }
+
+    /**
+     * Checks if an update is necessary, i.e., if [other] contains nodes which are not present in
+     * `this` and if the lattice of a node in [other] "is bigger" than the respective lattice in
+     * `this`. It does not modify anything.
+     */
+    open fun needsUpdate(other: State<K, V>): Boolean {
+        var update = false
+        for ((node, newLattice) in other) {
+            update = update || node !in this || newLattice > this[node]
+        }
+        return update
+    }
+
+    /** Deep copies this object. */
+    open fun duplicate(): State<K, V> {
+        val clone = State<K, V>()
+        for ((key, value) in this) {
+            clone[key] = value.duplicate()
+        }
+        return clone
+    }
+
+    /**
+     * Adds a new mapping from [newNode] to (a copy of) [newLattice] to this object if [newNode]
+     * does not exist in this state yet. If it already exists, it computes the least upper bound of
+     * [newLattice] and the current one for [newNode]. It returns if the state has changed.
+     */
+    open fun push(newNode: K, newLattice: Lattice<V>?): Boolean {
+        if (newLattice == null) {
+            return false
+        }
+        if (newNode in this && this[newNode]?.let { it >= newLattice } == true) {
+            // newLattice is "smaller" than the currently stored one. We don't add it anything.
+            return false
+        } else if (newNode in this) {
+            // newLattice is "bigger" than the currently stored one. We update it to the least
+            // upper bound
+            this[newNode] = newLattice.lub(this[newNode])
+        } else {
+            this[newNode] = newLattice.duplicate()
+        }
+        return true
+    }
+}
+
+/**
+ * A worklist. Essentially, it stores mappings of nodes to the states which are available there and
+ * determines which nodes have to be analyzed.
+ */
+class Worklist<K : Any, N : Any, V>() {
+    /** A mapping of nodes to the state which is currently available there. */
+    var globalState: MutableMap<K, State<N, V>> = mutableMapOf()
+        private set
+
+    /** A list of all nodes which have already been visited. */
+    private val alreadySeen = mutableListOf<K>()
```

Comment:
you probably want an identityhashset for the seen-list, otherwise you might run into performance problems if lists are large and equals/hashcode must be used.